### PR TITLE
perf(issues): getting users seen counts from cache rather than directly from Snuba

### PR DESCRIPTION
**Pull request by vartec**
This most basic is an optimization to not go to Snuba for _every_ new event in the issue. 
While the number of events seen since the last time we checked unique users count is less than then number of users needed to reach the threshold, we can be sure that the threshold was _not_ reached. 

Caveat: when the cached count is just one away from triggering, we'll still call Snuba on every new event in issue until new user is seen.

### CODE REVIEW

File: src/sentry/features/temporary.py [hunk 1]
```diff
@@ -80,6 +80,7 @@ def register_temporary_features(manager: FeatureManager):
     manager.add("organizations:grouping-suppress-unnecessary-secondary-hash", OrganizationFeature, FeatureHandlerStrategy.INTERNAL)
     manager.add("organizations:grouping-title-ui", OrganizationFeature, FeatureHandlerStrategy.REMOTE)
     manager.add("organizations:grouping-tree-ui", OrganizationFeature, FeatureHandlerStrategy.REMOTE)
+    manager.add("organizations:groupsnooze-cached-counts", OrganizationFeature, FeatureHandlerStrategy.OPTION)
```
Comment (wedamija) [inline src/sentry/features/temporary.py]: 🥳

File: src/sentry/models/groupsnooze.py [hunk 2]
```diff
@@ -142,6 +139,42 @@ def test_user_rates(self):
...
+        metrics.incr("groupsnooze.test_user_counts")
+        metrics.incr("groupsnooze.test_user_counts.real_count")
+
+        threshold = self.user_count + self.state["users_seen"]
+        real_count = group.count_users_seen(
+            referrer=Referrer.TAGSTORE_GET_GROUPS_USER_COUNTS_GROUP_SNOOZE.value
+        )
+        return real_count <= threshold
+
+    def test_user_counts_w_cache(self, group=None):
```
Comment (wedamija) [inline src/sentry/models/groupsnooze.py:160]: Is the plan to rename this after we remove the flag?

File: src/sentry/models/groupsnooze.py [hunk 3]
```diff
@@ -142,6 +139,42 @@ def test_user_rates(self):
...
+        metrics.incr("groupsnooze.test_user_counts")
+        cache_key = f"{self.get_cache_key(self.group_id)}:test_user_counts:events_seen_counter"
```
Comment (wedamija) [inline src/sentry/models/groupsnooze.py:162]: Actually wondering if this key should be the group snooze id. That way if the user deletes and recreates it we bust the cache automatically

File: src/sentry/models/groupsnooze.py [hunk 4]
```diff
@@ -88,13 +91,7 @@ def is_valid(self, group=None, test_rates=False, use_pending_data=False):
             if self.user_window:
                 if not self.test_user_rates():
```
Comment (wedamija) [inline src/sentry/models/groupsnooze.py:92]: Ahh I was confused, I guess we also need caching like this here, but that could be a follow up

File: src/sentry/models/groupsnooze.py [hunk 5]
```diff
@@ -142,6 +139,42 @@ def test_user_rates(self):
...
+        cache_key = f"{self.get_cache_key(self.group_id)}:test_user_counts:events_seen_counter"
+
+        threshold = self.user_count + self.state["users_seen"]
+
+        if cache.get(cache_key, float("inf")) < threshold:
+            # if we've seen less than that many events, we can't possibly have seen enough users
+            cache.increment(cache_key)
+            return True
+
+        metrics.incr("groupsnooze.test_user_counts.real_count")
```
Comment (wedamija) [inline src/sentry/models/groupsnooze.py:171]: Should we differentiate the metrics here so we can see the rates of real count change in `test_user_counts_w_cache` vs no cache?

**Review by wedamija — COMMENTED** [review]
I'm wondering if we should also clear the cache explicitly when we hit the threshold, and maybe if a user modifies the snooze?

File: src/sentry/models/groupsnooze.py (hunk 2 above)
Comment (vartec) [reply src/sentry/models/groupsnooze.py:160]: yeah, when we remove the flag the shim goes away and this one is renamed `test_user_counts`

File: src/sentry/models/groupsnooze.py (hunk 5 above)
Comment (vartec) [reply src/sentry/models/groupsnooze.py:171]: actually that should be a tag I guess

File: src/sentry/models/groupsnooze.py (hunk 3 above)
Comment (vartec) [reply src/sentry/models/groupsnooze.py:162]: makese sense, changed

**Comment by vartec** [comment]
> I'm wondering if we should also clear the cache explicitly when we hit the threshold, and maybe if a user modifies the snooze?

Is modifying a snooze actually modifying existing GroupSnooze or does it rather delete and recreate? 

In either case what's being kept in cache is the count of events seen since last time checked, so that's independent of GroupSnooze params.

File: src/sentry/models/groupsnooze.py [hunk 6]
```diff
@@ -60,6 +61,8 @@ class Meta:
 
     __repr__ = sane_repr("group_id")
 
+    USER_COUNT_CACHE_TIMEOUT = 300
```
Comment (anonrig) [inline src/sentry/models/groupsnooze.py:64]: Can you add a comment to here and explaining the unit of this variable? Is it `ms`?

File: src/sentry/models/groupsnooze.py [hunk 7]
```diff
@@ -142,6 +139,42 @@ def test_user_rates(self):
 
         return True
 
+    def test_user_counts(self, group=None):
```
Comment (anonrig) [inline src/sentry/models/groupsnooze.py:142]: Can you add comments?

File: src/sentry/models/groupsnooze.py [hunk 8]
```diff
@@ -142,6 +142,53 @@ def test_user_rates(self):
...
+
+    def test_user_counts_w_cache(self, group=None):
+        metrics.incr("groupsnooze.test_user_counts", tags={"cached": "true"})
+        cache_key = f"groupsnooze:v1:{self.id}:test_user_counts:events_seen_counter"
+
+        threshold = self.user_count + self.state["users_seen"]
+
+        if cache.get(cache_key, float("inf")) < threshold - 1:
+            # if we've seen less than that many events, we can't possibly have seen enough users
+            cache.incr(cache_key)
```
Comment (wedamija) [inline src/sentry/models/groupsnooze.py:182]: I think that we should also look at the post `incr` value that is returned, and if that's above `threshold-1` then we query. This just avoids a few race conditions where two processes fetch the same value, both increment, and one ends up over the threshold

**Review by wedamija — COMMENTED** [review]
lgtm, this approach makes sense

File: src/sentry/models/groupsnooze.py (hunk 8 above)
Comment (vartec) [reply src/sentry/models/groupsnooze.py]: good callout, I've changed it to increase and get value from that before the if check

**Review by wedamija — COMMENTED** [review]
I think this is good to go, just needs a couple tests

**Review by wedamija — APPROVED** [review]

**Pull request merged**